modules/apps/document-library/document-library-service/src/main/java/com/liferay/document/library/internal/configuration/admin/service/DLPreviewLimitManagedServiceFactory.java

```diff
@@ -0,0 +1,514 @@
+
+			throw new DLPreviewLimitConfigurationException.
+				InvalidMaxFileSizeException();
+		}
+	}
+
+	private void _checkMaxNumberOfPages(
+			int maxNumberOfPages, int maxNumberOfPagesLimit)
+		throws DLPreviewLimitConfigurationException.
+			InvalidMaxNumberOfPagesException {
+
+		if (_isLimitExceeded(
+				maxNumberOfPagesLimit, maxNumberOfPages,
+				_UNLIMITED_MAX_NUMBER_OF_PAGES)) {
+
+			throw new DLPreviewLimitConfigurationException.
+				InvalidMaxNumberOfPagesException();
+		}
+	}
+
+	private DLPreviewLimitConfiguration _getCompanyDLPreviewLimitConfiguration(
+		long companyId) {
+
+		return _getDLPreviewLimitConfiguration(
+			companyId, _companyConfigurationBeans,
+			() -> _systemDLPreviewLimitConfiguration);
+	}
+
+	private long _getCompanyMaxFileSize(long companyId) {
+		DLPreviewLimitConfiguration dlPreviewLimitConfiguration =
+			_getCompanyDLPreviewLimitConfiguration(companyId);
+
+		return dlPreviewLimitConfiguration.maxFileSize();
+	}
+
+	private int _getCompanyMaxNumberOfPages(long companyId) {
+		DLPreviewLimitConfiguration dlPreviewLimitConfiguration =
+			_getCompanyDLPreviewLimitConfiguration(companyId);
+
+		return dlPreviewLimitConfiguration.maxNumberOfPages();
+	}
+
+	private DLPreviewLimitConfiguration _getDLPreviewLimitConfiguration(
+		long key, Map<Long, DLPreviewLimitConfiguration> configurationBeans,
+		Supplier<DLPreviewLimitConfiguration> supplier) {
+
+		if (configurationBeans.containsKey(key)) {
+			return configurationBeans.get(key);
+		}
+
+		return supplier.get();
+	}
+
+	private DLPreviewLimitConfiguration _getGroupDLPreviewLimitConfiguration(
+		long groupId) {
+
+		return _getDLPreviewLimitConfiguration(
+			groupId, _groupConfigurationBeans,
+			() -> {
+				Group group = _groupLocalService.fetchGroup(groupId);
+
+				long companyId = CompanyThreadLocal.getCompanyId();
+
+				if (group != null) {
+					companyId = group.getCompanyId();
+				}
+
+				return _getCompanyDLPreviewLimitConfiguration(companyId);
+			});
+	}
+
+	private long _getGroupMaxFileSize(long groupId) {
+		DLPreviewLimitConfiguration dlPreviewLimitConfiguration =
+			_getGroupDLPreviewLimitConfiguration(groupId);
+
+		return dlPreviewLimitConfiguration.maxFileSize();
+	}
+
+	private int _getGroupMaxNumberOfPages(long groupId) {
+		DLPreviewLimitConfiguration dlPreviewLimitConfiguration =
+			_getGroupDLPreviewLimitConfiguration(groupId);
+
+		return dlPreviewLimitConfiguration.maxNumberOfPages();
+	}
+
+	private long _getMinimum(
+		long companyValue, long groupValue, long systemValue,
+		long unlimitedValue) {
+
+		if ((groupValue != unlimitedValue) &&
+			((systemValue == unlimitedValue) || (groupValue < systemValue)) &&
+			((companyValue == unlimitedValue) || (groupValue < companyValue))) {
+
+			return groupValue;
+		}
+
+		if ((companyValue != unlimitedValue) &&
+			((systemValue == unlimitedValue) || (companyValue < systemValue))) {
+
+			return companyValue;
+		}
+
+		return systemValue;
+	}
```

Comment:
This looks a bit hairy. Why don't you internally convert the unlimited value to Long.MAX_LONG and replace this method by Math.min?

Reply:
The `unlimitedValue` parameter references the value the user can set to make the settings option _unlimited_.

* Maximum file size: what's the maximum file size for the preview generation? `unlimitedValue = -1` means no limit
* Maximum number of pages: how many pages to generate in preview? `unlimitedValue = 0` means no limit

The method `_getMinimum` doesn't get the minimum value between long values. It get the most restrictive value in the settings hierarchy. E.g. group = 1, company = 2, system = 0 --> result = 1

Should I change the method name maybe to something like `_getMostRestrictive` ?

Reply:
What I meant to say is that, while keeping the -1 value as API, translate it internally to Long.MAX_VALUE so that the comparisons are simpler and the code is easier to understand. Then, once the comparisons are made, you translate it back to -1.

Reply:
Nice trick!

Reply:
I did the proposed refactoring but then I realized that this computation is meaningless. We already **enforce that a value in one scope is lower or equal to the higher level scope** in the UI, so if we check for the most restrictive value at a group level, the group value itself will be the most restrictive value because: group value <= company value <= system value.